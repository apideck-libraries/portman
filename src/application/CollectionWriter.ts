import { CollectionDefinition } from 'postman-collection'
import {
  injectEnvVariables,
  orderCollectionRequests,
  overwriteCollectionKeyValues,
  overwriteCollectionValues,
  overwriteCollectionSecurityValues,
  writeCollectionPreRequestScripts,
  writeRawReplacements
} from '.'
import { GlobalConfig, PortmanConfig, PortmanOptions } from '../types'
import { isEmptyObject } from 'utils'

export class CollectionWriter {
  public collection: CollectionDefinition
  config: PortmanConfig
  options: PortmanOptions

  constructor(config: PortmanConfig, options: PortmanOptions, collection: CollectionDefinition) {
    this.config = config
    this.options = options
    this.collection = collection
  }

  public execute(): void {
    const { envFile, baseUrl } = this.options

    const { globals = {} } = this.config
    const {
      collectionPreRequestScripts = [],
      securityOverwrites = {},
      keyValueReplacements = {},
      valueReplacements = {},
      rawReplacements = [],
      orderOfOperations = []
    } = globals as GlobalConfig

    let collection = this.collection

    // --- Portman - Set Security values for Postman
    if (securityOverwrites && !isEmptyObject(securityOverwrites)) {
      collection = overwriteCollectionSecurityValues(collection, securityOverwrites)
    }

    // --- Portman - Search for keys in dictionary to set the value if key is found anywhere in collection
    if (keyValueReplacements && !isEmptyObject(keyValueReplacements)) {
      collection = overwriteCollectionKeyValues(collection, keyValueReplacements)
    }

    // --- Portman - Search for keys in dictionary to set the values if value is found anywhere in collection
    if (valueReplacements && !isEmptyObject(valueReplacements)) {
      collection = overwriteCollectionValues(collection, valueReplacements)
    }

    // --- Portman - Use .env to inject variables on Collection
    if (envFile) {
      collection = injectEnvVariables(collection, envFile, baseUrl)
    }

    // --- Portman - Set manually order Postman requests
    if (orderOfOperations && orderOfOperations.length > 0) {
      collection = orderCollectionRequests(collection, orderOfOperations)
    }

    // --- Portman - Set Postman pre-requests
    if (collectionPreRequestScripts && collectionPreRequestScripts.length > 0) {
      collection = writeCollectionPreRequestScripts(collection, collectionPreRequestScripts)
    }

    // --- Portman - Replace & clean-up Postman
    if (rawReplacements && rawReplacements.length > 0) {
      const collectionString = writeRawReplacements(
        JSON.stringify(collection, null, 2),
        rawReplacements
      )
      this.collection = JSON.parse(collectionString)
    } else {
      this.collection = collection
    }
  }
}
