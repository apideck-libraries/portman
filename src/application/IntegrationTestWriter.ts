import { Collection, Item, ItemGroup } from 'postman-collection'
import { IntegrationTestConfig, VariationConfig, PortmanTestTypes } from '../types'
import { TestSuite, VariationWriter } from './'

export class IntegrationTestWriter {
  public testSuite: TestSuite
  integrationTestFolder: ItemGroup<Item>
  integrationTestCollection: Collection

  constructor({
    testSuite,
    integrationTestFolderName
  }: {
    testSuite: TestSuite
    integrationTestFolderName: string
  }) {
    this.testSuite = testSuite
    this.integrationTestFolder = new ItemGroup<Item>({
      name: integrationTestFolderName
    })
    this.integrationTestCollection = new Collection()
  }

  public add(integrationTest: IntegrationTestConfig): void {
    const { testSuite } = this
    const { name, operations } = integrationTest

    const variationWriter = new VariationWriter({
      testSuite: this.testSuite,
      variationFolderName: name
    })

    operations.map(operation => {
      const variations = (operation?.variations || []) as VariationConfig[]
      const { openApiOperationId } = operation

      const pmOperation = testSuite.postmanParser.getOperationById(openApiOperationId)

      if (!pmOperation) return

      const oaOperation = testSuite.oasParser.getOperationByPath(pmOperation.pathRef)

      // Handle integration variations
      variations.map(variation => {
        const variationName = variation.name

        const operationVariation = pmOperation.clone({
          newId: `${variationName}-${Math.random()}`,
          name: variationName
        })

        // Set/Update Portman operation test type
        this.testSuite.registerOperationTestType(operationVariation, PortmanTestTypes.integration)

        variationWriter.injectVariations(operationVariation, oaOperation, variation, operation)
        variationWriter.addToFolder(operationVariation, variationWriter.variationFolder)
      })
    })

    this.integrationTestCollection = variationWriter.mergeToCollection(
      this.integrationTestCollection
    )
  }

  public mergeToCollection(collection: Collection): Collection {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this.integrationTestCollection.items.map((item: any) => {
      this.integrationTestFolder.items.add(item)
    })

    collection.items.add(this.integrationTestFolder)
    return collection
  }
}
