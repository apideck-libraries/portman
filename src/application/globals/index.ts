export * from './injectEnvVariables'
export * from './orderCollectionRequests'
export * from './overwriteCollectionKeyValues'
export * from './overwriteCollectionSecurityValues'
export * from './overwriteCollectionValues'
export * from './renamePostmanCollection'
export * from './stripResponseExamples'
export * from './writeCollectionPreRequestScripts'
export * from './writeRawReplacements'
