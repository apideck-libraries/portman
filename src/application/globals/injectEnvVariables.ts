import { camelCase } from 'camel-case'
import { config } from 'dotenv'
import path from 'path'
import { CollectionDefinition, VariableDefinition } from 'postman-collection'

const upsertEnvVariable = (
  variables: VariableDefinition[],
  key: string,
  value: unknown,
  type: string
) => {
  const toInject = { key, value, type }
  const updateIndex = variables.findIndex(variable => variable['key'] === key)

  if (updateIndex > -1) {
    variables[updateIndex] = toInject
  } else {
    variables.push(toInject)
  }

  return variables
}

export const injectEnvVariables = (
  obj: CollectionDefinition,
  envFile: string | undefined,
  baseUrl: string | undefined
): CollectionDefinition => {
  let variables = (obj.variable as VariableDefinition[]) || []
  const baseUrlFromSpec = variables.find(item => {
    return item.id === 'baseUrl'
  })?.value

  envFile && config({ path: path.resolve(envFile) })
  const parsed = process.env

  for (const [key, val] of Object.entries(parsed)) {
    if (key.startsWith('PORTMAN_')) {
      const id = key.replace('PORTMAN_', '')
      variables = upsertEnvVariable(variables, camelCase(id), val, 'string')
    }
  }

  if (baseUrl || baseUrlFromSpec) {
    variables = upsertEnvVariable(variables, 'baseUrl', baseUrl || baseUrlFromSpec, 'string')
  }
  const uniqueVariables = Array.from(new Set(variables))

  const collection = JSON.parse(JSON.stringify(obj))
  collection.variable = uniqueVariables

  return collection as CollectionDefinition
}
