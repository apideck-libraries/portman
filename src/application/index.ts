export * from './CollectionWriter'
export * from './globals'
export * from './newman'
export * from './overwrites'
export * from './tests'
export * from './TestSuite'
export * from './variables'
