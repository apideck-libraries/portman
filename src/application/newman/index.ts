export * from './runNewmanWith'
export * from './writeNewmanEnv'
