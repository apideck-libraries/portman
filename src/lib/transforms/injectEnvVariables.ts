import { camelCase } from 'camel-case'
import { config } from 'dotenv'
import path from 'path'
import { CollectionDefinition, VariableDefinition } from 'postman-collection'

const upsertEnvVariable = (
  variables: VariableDefinition[],
  key: string,
  value: unknown,
  type: string
) => {
  const toInject = { key, value, type }
  const updateIndex = variables.findIndex(variable => variable['id'] === key)

  if (updateIndex > -1) {
    variables[updateIndex] = toInject
  } else {
    variables.push(toInject)
  }

  return variables
}

export const injectEnvVariables = (
  obj: CollectionDefinition,
  envFile: string | undefined,
  baseUrl: string | undefined
): CollectionDefinition => {
  if (!envFile) return obj

  let variables = (obj.variable as VariableDefinition[]) || []
  const baseUrlFromSpec = variables.find(item => {
    return (item.id = 'baseUrl')
  })?.value

  const { parsed = {} } = config({ path: path.resolve(envFile) })

  for (const [key, val] of Object.entries(parsed)) {
    if (key.startsWith('PORTMAN_')) {
      const id = key.replace('PORTMAN_', '')
      variables = upsertEnvVariable(variables, camelCase(id), val, 'string')
    }
  }

  variables = upsertEnvVariable(variables, 'baseUrl', baseUrl || baseUrlFromSpec, 'string')
  obj.variable = [...new Set(variables)]
  return obj
}
