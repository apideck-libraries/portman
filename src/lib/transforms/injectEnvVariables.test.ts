import fs from 'fs-extra'
import { OpenApiParser, PostmanParser, TestSuiteService } from '../../application'
import { injectEnvVariables } from './injectEnvVariables'

describe('injectEnvVariables', () => {
  let postmanParser: PostmanParser
  let oasParser: OpenApiParser
  let testSuiteService: TestSuiteService

  const postmanJson = '__tests__/fixtures/crm.postman.json'
  const oasYml = '__tests__/fixtures/crm.yml'
  const testSuiteConfigFile = '__tests__/fixtures/postman-testsuite.crm.json'
  const envFile = '__tests__/fixtures/.crm.env'

  beforeEach(async () => {
    oasParser = new OpenApiParser()
    await oasParser.convert({ inputFile: oasYml })
    const postmanObj = JSON.parse(fs.readFileSync(postmanJson).toString())
    postmanParser = new PostmanParser({ postmanObj: postmanObj, oasParser: oasParser })
    testSuiteService = new TestSuiteService({ oasParser, postmanParser, testSuiteConfigFile })
    testSuiteService.generateAutomatedTests()
  })

  it('should add script to skip tests if status returned is Not Implemented', async () => {
    const collection = injectEnvVariables(testSuiteService.collection.toJSON(), envFile, undefined)
    expect(collection?.variable).toMatchSnapshot()
  })
})
