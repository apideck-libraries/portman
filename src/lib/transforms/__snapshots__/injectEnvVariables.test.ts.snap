// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectEnvVariables should add script to skip tests if status returned is Not Implemented 1`] = `
Array [
  Object {
    "id": "baseUrl",
    "type": "string",
    "value": "https://unify.apideck.com",
  },
  Object {
    "key": "bearerToken",
    "type": "string",
    "value": "YOU_API_KEY_HERE",
  },
  Object {
    "key": "applicationId",
    "type": "string",
    "value": "1111",
  },
  Object {
    "key": "serviceId",
    "type": "string",
    "value": "teamleader",
  },
  Object {
    "key": "consumerId",
    "type": "string",
    "value": "test_user_id",
  },
  Object {
    "id": "bearerToken",
    "type": "string",
    "value": "f7a82583-e42a-47ec-8325-f9e5ef231303-KAwRX5cg3DgAhvz5BKs8-9200e88f-b82e-4980-99ef-79aabd71ee48",
  },
  Object {
    "id": "unifiedApi",
    "type": "string",
    "value": "crm",
  },
  Object {
    "id": "applicationId",
    "type": "string",
    "value": "2222",
  },
  Object {
    "id": "serviceId",
    "type": "string",
    "value": "activecampaign",
  },
  Object {
    "id": "consumerId",
    "type": "string",
    "value": "testing_user_id",
  },
  Object {
    "id": "resource",
    "type": "string",
    "value": "lead",
  },
  Object {
    "id": "connectorApiKey",
    "type": "string",
    "value": "a1ffa05a296f99780689a6ebd44b7c966e5591ee13cd38e5116c26e22104ac1bfaa9e00a",
  },
]
`;
